src/fairchem/core/units/mlip_unit/inference/inference_server_ray.py

```diff
@@ -0,0 +1,148 @@
+    assign_device_for_local_rank,
+    get_device_for_local_rank,
+    setup_env_local_multi_gpu,
+)
+
+logging.basicConfig(level=logging.INFO)
+
+
+@ray.remote
+class MLIPWorker:
+    def __init__(
+        self, worker_id: int, world_size: int, master_port: int, predictor_config: dict
+    ):
+        self.worker_id = worker_id
+        self._distributed_setup(
+            worker_id, master_port, world_size, predictor_config.get("device", "cpu")
+        )
+        self.predict_unit = hydra.utils.instantiate(predictor_config)
+        logging.info(
+            f"Worker {worker_id}, gpu_id: {ray.get_gpu_ids()}, loaded predict unit: {self.predict_unit}, "
+            f"on port {master_port}, with device: {get_device_for_local_rank()}, config: {predictor_config}"
+        )
+
+    def _distributed_setup(
+        self, worker_id: int, master_port: int, world_size: int, device: str
+    ):
+        # initialize distributed environment
+        # TODO, this wont work for multi-node, need to fix master addr
+        setup_env_local_multi_gpu(worker_id, master_port)
+        # local_rank = int(os.environ["LOCAL_RANK"])
+        assign_device_for_local_rank(device == "cpu", 0)
+        backend = "gloo" if device == "cpu" else "nccl"
+        dist.init_process_group(
+            backend=backend,
+            rank=worker_id,
+            world_size=world_size,
+        )
+        gp_utils.setup_graph_parallel_groups(world_size, backend)
+
+    def predict(self, data: bytes):
+        atomic_data = pickle.loads(data)
+        result = self.predict_unit.predict(atomic_data)
+        return pickle.dumps(result)
+
+
+class MLIPInferenceServerWebSocket:
+    def __init__(self, predictor_config: dict, port=8001, num_workers=1):
+        logging.basicConfig(level=logging.INFO)
+        self.host = "localhost"
+        self.port = port
+        self.num_workers = num_workers
+        self.predictor_config = predictor_config
+        # Initialize a pool of MLIPWorkers
+        self.master_pg_port = get_free_port()
+        ray.init(logging_level=logging.INFO)
+        options = {"num_gpus": 1} if predictor_config.get("device") == "cuda" else {}
+        self.workers = [
+            MLIPWorker.options(**options).remote(
+                i, self.num_workers, self.master_pg_port, self.predictor_config
+            )
+            for i in range(self.num_workers)
+        ]
+        logging.info(
+            "Initialized Local MLIPInferenceServerWebSocket with config: "
+            f"{self.predictor_config}, port: {self.port}, workers: {self.num_workers}"
+        )
+
+    async def handler(self, websocket):
+        try:
+            async for message in websocket:
+                # don't unpickle here, just pass bytes to workers
+                futures = [w.predict.remote(message) for w in self.workers]
+                # only need to retrieve results from the first worker since they are identical
+                results = ray.get(futures[0])
```

Comment:
Do these futures pile up anywhere on the worker (worker_id 1+). side? do they need to be received?